codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/KotlinClientSerdeDecorator.kt

```diff
@@ -0,0 +1,654 @@
+                                "serializer.serialize_newtype_variant(${unionName.dq()}, $index, $fieldName, #{member})",
+                                "member" to serializeMember(member, "inner"),
+                            )
+                        }
+                    }
+                    if (codegenContext.target.renderUnknownVariant()) {
+                        rustTemplate(
+                            "#{Union}::${UnionGenerator.UNKNOWN_VARIANT_NAME} => serializer.serialize_str(\"unknown variant!\")",
+                            "Union" to unionSymbol,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    private fun serializeDateTime(shape: TimestampShape): RuntimeType =
+        RuntimeType.forInlineFun("SerializeDateTime", Module) {
+            implSerializeConfigured(codegenContext.symbolProvider.toSymbol(shape)) {
+                rust("serializer.serialize_str(&self.value.to_string())")
+            }
+        }
+
+    private fun serializeDocument(shape: DocumentShape): RuntimeType =
+        RuntimeType.forInlineFun("SerializeDocument", Module) {
+            implSerializeConfigured(codegenContext.symbolProvider.toSymbol(shape)) {
+                rustTemplate(
+                    """
+                    use #{SerializeConfigured};
+                    match self.value {
+                         #{Document}::String(v) => serializer.serialize_str(v),
+                         #{Document}::Object(v) => {
+                             use #{serde}::ser::SerializeMap;
+                             let mut map = serializer.serialize_map(Some(v.len()))?;
+                             for (k, v) in v.iter() {
+                                 map.serialize_entry(k, &v.serialize_ref(&self.settings))?;
+                             }
+                             map.end()
+                         },
+                         #{Document}::Array(v) => {
+                             use #{serde}::ser::SerializeSeq;
+                             let mut seq = serializer.serialize_seq(Some(v.len()))?;
+                             for e in v.iter() {
+                                 seq.serialize_element(&e.serialize_ref(&self.settings))?;
+                             }
+                             seq.end()
+                         },
+                         #{Document}::Number(#{Number}::Float(value)) => value.serialize(serializer),
+                         #{Document}::Number(#{Number}::PosInt(value)) => {
+                             value.serialize(serializer)
+                         },
+                         #{Document}::Number(#{Number}::NegInt(value)) => {
+                             value.serialize(serializer)
+                         },
+                         #{Document}::Bool(b) => b.serialize(serializer),
+                         #{Document}::Null => serializer.serialize_none(),
+                     }
```

Comment:
Could this be implemented as part of a runtime crate so that we generate less code?

Reply:
No. `SerializeConfigured` is crate local (and I think it needs to be for coherence). I can experiment with making it _not_ crate local but I would probably start by defining it in the crate.

The total amount of generated code is only dozens of lines which is totally dwarfed by the SDK code though, so I don't think there is a huge impact.

Reply:
We _could_ put it in an inlinable if you prefer though.